tests/test_pytypes.cpp

```diff
@@ -462,6 +462,34 @@ TEST_SUBMODULE(pytypes, m) {
     m.def("weakref_from_object_and_function",
           [](py::object o, py::function f) { return py::weakref(std::move(o), std::move(f)); });
 
+    m.def("tuple_iterator", []() {
+        auto tup = py::make_tuple(5, 7);
+        int tup_sum = 0;
+        for (auto &it : tup) {
+            tup_sum += it.cast<int>();
+        }
+        return tup_sum;
+    });
+
+    m.def("dict_iterator", []() {
+        py::dict dct;
+        dct[py::int_(3)] = 5;
```

Comment:
Okay, so this is a bug with storing const values in keys?

Reply:
I'm not sure I understand the question, but I'm thinking the answer is no.
This code here is just to have a couple random dict entries. The actual lines that matter are only the three `for (auto &it : xxx)`.
IIUC, a new `handle` is constructed in each iteration, a temporary. `auto &it` creates a reference to a temporary. The stricter compilers that reject this code even if we keep the `const` are doing the "right thing" IMO.
For us the critical question is: do we want to make pybind11 a forcing function (forcing clean user code by removing the `const`)? On that one I'm ambivalent, leaning to no, to not have to deal with the fallout (is that lazy or smart?).
The next question: what do we want to do about these tests here?
If we remove them from this PR, it will not be obvious at all why we need the NOLINTs.
If we keep them, we need ifdefs do make them work with the stricter compilers.
Hm ... writing this I'm beginning to think we just need to insert `const` for those strict compilers. It'll look a little convoluted, but then we have all situations covered in code, as concise as it can be made.